### PR TITLE
Add Rust build support [Rebase & FF]

### PR DESCRIPTION
**Pull request by makubacki**
## Description

Adds Rust build support to Project Mu.

### Overview

These changes enable a platform to build Rust code as part of the normal firmware build process.

[Cargo make](https://crates.io/crates/cargo-make/) is used as a build runner to allow users to build Rust packages individually and simply on the command line similar to the way they are built during the firmware build process. It is a CLI tool that helps abstract away many of the CLI arguments necessary to build a Rust module so that developers can easily check, test, and build individual Rust packages without the need to copy and/or memorize a long list of arguments.

Includes changes from: https://github.com/tianocore/edk2-staging/tree/edkii-rust

### Generally Getting Started with Rust

It is recommended to:

1. Update `mu_basecore` to include these changes

2. Download and install `Rust` and `cargo` from [Getting Started - Rust Programming Language (rust-lang.org)](https://www.rust-lang.org/learn/get-started)

3. Verify `cargo` is working:

   \>`cargo --version`

4. Install the desired Rust tool chain

   - Example: `1.68.2 x86_64 toolchain`

   - Windows:

      \>`rustup toolchain install 1.68.2-x86_64-pc-windows-msvc`

      \>`rustup component add rust-src --toolchain 1.68.2-x86_64-pc-windows-msvc`

   - Linux:

      \>`rustup toolchain install 1.68.2-x86_64-unknown-linux-gnu`

      \>`rustup component add rust-src --toolchain 1.68.2-x86_64-unknown-linux-gnu`

5. Install `cargo make`

   \>`cargo install --force cargo-make`

At this point, the essential Rust applications are installed, and a repo can begin to add and build Rust code.

### Build Variations

First, the simplest case is a new user approaching a package that supports Rust modules. In that case, just build the package like "normal" (e.g., stuart build) and the "normal" build output (i.e., EFI drivers and firmware volumes) will be produced.

#### Build an Individual Module

  ```cmd
  cargo make build <Module Name>
  ```

The following command line options are available:

  1. `-p PROFILE [development|release]`. `DEFAULT` = `development (debug)`
  2. `-e ARCH=[IA32|X64|LOCAL]`. `DEFAULT` = `X64`
  3. `-e TARGET_TRIPLE=[triple]`.

  - `ARCH=LOCAL` is used to build any locally executable tools associated with a Rust library package (e.g., a dual-purpose executable and library).

  - `TARGET_TRIPLE=<triple>` is used to cross-compile locally executable tools associated with a Rust library package.

  - Note: A Rust package must be specified.
  - The output location is:
    - `target/[x86_64-unknown-uefi|i686-unknown-uefi]/[debug|release]/module_name.[efi|rlib]`

#### Test an Individual Module

  ```cmd
  cargo make test <Optional: Module Name>
  ```

  - Note: If a package is not specified, all packages will be tested.

#### Supported Build Combinations

1. C source + Rust source mixed in INF (Library or Module)
   - Rust source code is supported by build rule – `Rust-To-Lib-File` (`.rs` => `.lib`)

   - Limitation: Rust code cannot have external dependencies.

2. Pure Rust Module only.

   - A `Cargo.toml` file is added to the INF file `[Sources]` section.

   - Rust Module build is supported by build rule – `Toml-File.RUST_MODULE` (`Toml` => `.efi`)

   - Limitation: Runtime might be a problem, such as virtual address translation for Rust code internal global variables.

3. Pure Rust Module + Pure Rust Library with Cargo Dependency.

   - Cargo dependency means the Rust lib dependency declared in `Cargo.toml`.

4. Pure Rust Module + C Library with EDK II Dependency.

   - Rust Module build is supported by build rule – `Toml-File` (`Toml` => `.lib`)

   - The EDK II dependency means the EDK II lib dependency declared in INF.

     - If a Rust module is built with C, the cargo must use `staticlib`. Otherwise, `rlib` is used.

5. C Module + Pure Rust Library with EDK II Dependency.

   - Rust library build is supported by build rule – `Toml-File`. (`Toml` => `.lib`)

6. Pure Rust Module + Pure Rust Library with EDK II Dependency.
   - Same as (4) + (5).

- [ ] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [ ] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [ ] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

1. Verified build of packages with no Rust code
2. Verified build of packages with Rust code in various combinations:
   - Pure Rust module
   - Pure Rust module with Cargo dependency
   - Pure Rust module with C (EDK II) library

## Integration Instructions

Included and PR description and markdown files in the PR.

### CODE REVIEW

**Comment by Javagedes** [comment]
@makubacki, Do you want to grab cargo-make directly from https://github.com/sagiegurari/cargo-make as an external dependency as we do with basetools? If so, I can add it in a PR after this.

**Comment by makubacki** [comment]
> @makubacki, Do you want to grab cargo-make directly from https://github.com/sagiegurari/cargo-make as an external dependency as we do with basetools? If so, I can add it in a PR after this.

It's something to consider. I didn't view the current installation method as too much trouble though. Personally, I'd prefer to wait and see how that fits scaling out when more repos add Rust. There are some plugins for testing and code coverage I was thinking should get added after this so they're more broadly available.

File: Docs/rust_build.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+   - Cargo dependency means the Rust lib dependency declared in `Cargo.toml`.
+
+4. Pure Rust Module + C Library with EDK II Dependency.
+
+   - Rust Module build is supported by build rule – `Toml-File` (`Toml` => `.lib`)
+
+   - The EDK II dependency means the EDK II lib dependency declared in INF.
+
+     - If a Rust module is built with C, the cargo must use `staticlib`. Otherwise, `rlib` is used.
```
Comment (Javagedes) [inline Docs/rust_build.md:125]: I think this should be a bit more exact. i.e. specifically call out (or link to rust docs https://doc.rust-lang.org/cargo/reference/cargo-targets.html#library) that they need to set crate-type = ['staticlib'] at the minimum under the [lib] section.

Comment (makubacki) [reply Docs/rust_build.md:125]: Added

**Comment by apop5** [comment]
Only only x86 being targeted here? Would now be a good time to add aarch64? Or should we do that I the future?

**Review by apop5 — APPROVED** [review]

**Comment by Javagedes** [comment]
> Only only x86 being targeted here? Would now be a good time to add aarch64? Or should we do that I the future?

aarch64-unknown-uefi is a supported build target by rust so it won't be difficult to add in @makubacki I can add it to this PR or a follow up PR. Let me know.

**Comment by makubacki** [comment]
> aarch64-unknown-uefi is a supported build target by rust so it won't be difficult to add in @makubacki I can add it to this PR or a follow up PR. Let me know.

We're really only targeting x86/64 development right now. I'm mostly indifferent to adding generic support now versus waiting until we have an actual need for AARCH64 and can document and test it regularly. @Javagedes if you'd like to add it to the PR, that's fine. Just let me know.

**Comment by Javagedes** [comment]
> > aarch64-unknown-uefi is a supported build target by rust so it won't be difficult to add in @makubacki I can add it to this PR or a follow up PR. Let me know.
> 
> We're really only targeting x86/64 development right now. I'm mostly indifferent to adding generic support now versus waiting until we have an actual need for AARCH64 and can document and test it regularly. @Javagedes if you'd like to add it to the PR, that's fine. Just let me know.

I'll come back with another PR when I get some free time.

**Review by Javagedes — APPROVED** [review]

**Pull request merged**